sdks/go/pkg/beam/runners/prism/internal/engine/strategy.go

```diff
@@ -23,16 +23,552 @@ import (
+	} else {
+		t.Main.onFire(state)
+		if state.getTriggerState(t.Main).finished {
+			t.Main.reset(state)
+		}
+	}
+}
+
+func (t *TriggerOrFinally) reset(state *StateData) {
+	t.Main.reset(state)
+	t.Finally.reset(state)
+	delete(state.Trigger, t)
+}
+
+func (t *TriggerOrFinally) String() string {
+	return fmt.Sprintf("OrFinally[Repeat:%v Until:%v]", t.Main, t.Finally)
+}
+
+// TriggerRepeatedly is a composite trigger that will fire whenever the Repeated trigger is ready.
+// If the Repeated trigger is finished, it's state will be reset.
+type TriggerRepeatedly struct {
+	Repeated Trigger
+}
+
+func (t *TriggerRepeatedly) onElement(input triggerInput, state *StateData) {
+	t.Repeated.onElement(input, state)
+}
+
+func (t *TriggerRepeatedly) shouldFire(state *StateData) bool {
+	return t.Repeated.shouldFire(state)
+}
+
+func (t *TriggerRepeatedly) onFire(state *StateData) {
+	if !t.shouldFire(state) {
+		return
+	}
+	t.Repeated.onFire(state)
+	// If the subtrigger is finished, reset it.
+	if repeatedTs := state.getTriggerState(t.Repeated); repeatedTs.finished {
+		t.Repeated.reset(state)
+	}
+}
+
+func (t *TriggerRepeatedly) reset(state *StateData) {
+	t.Repeated.reset(state)
+	delete(state.Trigger, t)
+}
+
+func (t *TriggerRepeatedly) String() string {
+	return fmt.Sprintf("Repeat[%v]", t.Repeated)
+}
+
+// TriggerAfterEndOfWindow is a composite trigger that will fire whenever the
+// the early Triggers are ready prior to the end of window, implicitly repeated.
+// After the end of window, the Late trigger will be implicitly repeated.
+//
+// Uses the extra state field to track if the end of the window has been reached.
+type TriggerAfterEndOfWindow struct {
+	Early, Late Trigger
+}
+
+func (t *TriggerAfterEndOfWindow) onElement(input triggerInput, state *StateData) {
+	ts := state.getTriggerState(t)
+	if ts.finished {
+		return
+	}
+	if ts.extra == nil {
+		ts.extra = false
+	}
+	previouslyEndOfWindow := ts.extra.(bool)
+	if !previouslyEndOfWindow && input.endOfWindowReached {
+		// We have transitioned. Clear early state and mark it finished
+		triggerClearAndFinish(t.Early, state)
+		if t.Late == nil {
+			triggerClearAndFinish(t, state)
+			return
+		}
+	}
+	ts.extra = input.endOfWindowReached
+	state.setTriggerState(t, ts)
+
+	if !state.getTriggerState(t.Early).finished {
+		t.Early.onElement(input, state)
+		return
+	} else if t.Late != nil {
+		t.Late.onElement(input, state)
+	}
+}
+
+func (t *TriggerAfterEndOfWindow) shouldFire(state *StateData) bool {
+	ts := state.getTriggerState(t)
+	if ts.finished {
+		return false
+	}
+	if !state.getTriggerState(t.Early).finished {
+		return t.Early.shouldFire(state) || ts.extra.(bool)
+	} else if t.Late == nil {
+		return false
+	}
+	return t.Late.shouldFire(state)
+}
+
+func (t *TriggerAfterEndOfWindow) onFire(state *StateData) {
+	ts := state.getTriggerState(t)
+	if ts.finished {
+		return
+	}
+	if !state.getTriggerState(t.Early).finished {
+		if t.Early.shouldFire(state) {
```

Comment:
Should we check `ts.extra` here as well? I don't really follow how `ts.extra` is being used in this context

Reply:
I had that in a previous version but it was leading to incorrect testing behaviors:

In particular, in the Java implementation, it uses triggers for *all* firings, Early, and Late, but also OnTime and Closing (which is either OnTime or Late).

The implementation I've settled on is not to do that, since OnTime and Closing are already handled and only occur when the watermark has advanced anyway.

So the extra state is mostly there to mark that we transition from the Early firings trigger (if any) to the Late firings trigger (if any).

Reply:
What you describe could cause weirdness for a corner case that is rare but I got called out for: if someone has windowing but also the trigger is `Repeat(ElementCount(n)` or `Repeat(ProcessingTime(30s))` then they are not expecting an extra firing when the watermark crosses the end of the window. Presumably Prism would emit at that point. I honestly doubt any reasonable pipeline would notice the difference, but there you have it. (note that for a trigger `Repeat(ProcessingTime(30s))` we still mark the final non-late pane as "on time")

In the Java implementation the end-of-window firing is still governed by the trigger (but _also_ governed by `OnTimeBehavior`), but the expiry firing is governed only by `ClosingBehavior`.

Reply:
At present, in the follow up PR, Prism will always Emit at the end of window, except for the never trigger, which would be only at the closing time.

It doesn't make sense to me for triggers to be designed or implemented in such a way to lead to a trivial method of dataloss. That is against the point of a data processing system.

Is there a benefit to allow for trivial dataloss, vs having the sensible behavior of always firing at the end of the window?